src/storage/src/window_infer.rs

```diff
@@ -0,0 +1,328 @@
+        -> Vec<TimestampRange>;
+}
+
+/// [PlainWindowInference] simply finds the minimum time span within all SST files in level 0 and
+/// memtables, matches that time span into a set of predefined time windows.
+pub(crate) struct PlainWindowInference;
+
+impl WindowInfer for PlainWindowInference {
+    fn infer_window(
+        &self,
+        files: &[FileMeta],
+        mem_tables: &[MemtableStats],
+    ) -> Vec<TimestampRange> {
+        let mut min_duration_sec = i64::MAX;
+        let mut durations = Vec::with_capacity(files.len() + mem_tables.len());
+
+        for meta in files {
+            if let Some((start, end)) = &meta.time_range {
+                // unwrap safety: converting timestamps with any unit to seconds won't overflow.
+                let start_sec = start.convert_to(TimeUnit::Second).unwrap().value();
+                // file timestamp range end is inclusive
+                let end_sec = end.convert_to_ceil(TimeUnit::Second).unwrap().value();
+                debug_assert!(end_sec >= start_sec);
+                if meta.level == 0 {
+                    // only level 0 is involved when calculating time windows.
+                    min_duration_sec = min_duration_sec.min(end_sec - start_sec);
+                }
+                durations.push((start_sec, end_sec));
+            }
+        }
+
+        for stats in mem_tables {
+            // unwrap safety: converting timestamps with any unit to seconds won't overflow.
+            let start_sec = stats
+                .min_timestamp
+                .convert_to(TimeUnit::Second)
+                .unwrap()
+                .value();
+            let end_sec = stats
+                .max_timestamp
+                .convert_to_ceil(TimeUnit::Second)
+                .unwrap()
+                .value();
+            min_duration_sec = min_duration_sec.min(end_sec - start_sec);
+            durations.push((start_sec, end_sec));
+        }
+
+        let window_size = min_duration_to_window_size(min_duration_sec);
+        align_durations_to_windows(&durations, window_size)
+            .into_iter()
+            .sorted_by(|(l_start, _), (r_start, _)| r_start.cmp(l_start)) // sort time windows in descending order
+            // unwrap safety: we ensure that end>=start so that TimestampRange::with_unit won't return None
+            .map(|(start, end)| TimestampRange::with_unit(start, end, TimeUnit::Second).unwrap())
+            .collect()
+    }
+}
+
+/// Given a set of time spans and a min duration, this function aligns the time spans to windows that
+/// collectively covers all the time spans.
+///
+/// For example, given time span `[1, 6)` and duration 5, the span can be aligned and split to
+/// two windows with length 5: `[0, 5)` and `[5, 10]`, and these two windows can cover the original
+/// span `[1, 6)`.
+fn align_durations_to_windows(durations: &[(i64, i64)], min_duration: i64) -> HashSet<(i64, i64)> {
```

Comment:
nit: "duration" usually refers to a relative range of time. How about `align_time_spans_to_windows`?